Source/WebCore/Modules/url-pattern/URLPatternParser.cpp

```diff
@@ -330,5 +329,219 @@ String escapeRegexString(StringView input)
+                needsGrouping = isValidNameCodepoint(*StringView(nextPart->value).codePoints().begin(), true);
+            } else
+                needsGrouping = isASCIIDigit(nextPart->name[0]);
+        }
+
+        if (!needsGrouping && part.prefix.isEmpty() && previousPart && previousPart->type == PartType::FixedText) {
+            if (options.prefixCodepoint.length() == 1
+                && options.prefixCodepoint.startsWith(*StringView(nextPart->value).codePoints().codePointAt(nextPart->value.length() - 1)))
+                needsGrouping = true;
+        }
+
+        ASSERT(!part.name.isEmpty());
+
+        if (needsGrouping)
+            result.append("{"_s);
+
+        result.append(escapePatternString(part.prefix));
+
+        if (customName)
+            result.append(makeString(":"_s, part.name));
+
+        if (part.type == PartType::Regexp)
+            result.append("("_s, part.value, ")"_s);
+        else if (part.type == PartType::SegmentWildcard && !customName)
+            result.append(makeString("("_s, generateSegmentWildcardRegexp(options), ")"_s));
+        else if (part.type == PartType::FullWildcard) {
+            if (!customName
+                && (!previousPart || previousPart->type == PartType::FixedText || previousPart->modifier != Modifier::None
+                    || needsGrouping || !part.prefix.isEmpty()))
+                result.append("*"_s);
+            else
+                result.append(makeString("("_s, ".*"_s,  ")"_s));
+        }
+
+        if (part.type == PartType::SegmentWildcard && customName && !part.suffix.isEmpty() && isValidNameCodepoint(*StringView(part.suffix).codePoints().begin(), true))
+            result.append("\\"_s);
+
+        result.append(escapePatternString(part.suffix));
+
+        if (needsGrouping)
+            result.append("}"_s);
+
+        result.append(convertModifierToString(part.modifier));
+    }
+
+    return result.toString();
+}
+
+template<typename CharacterType>
+static String escapePatternStringForCharacters(std::span<const CharacterType> characters)
+{
+    static constexpr std::array escapeCharacters { '+', '*', '?', ':', '(', ')', '\\', '{', '}' }; // NOLINT
+
+    StringBuilder result;
+    result.reserveCapacity(characters.size());
+
+    for (auto character : characters) {
+        if (std::find(escapeCharacters.begin(), escapeCharacters.end(), character) != escapeCharacters.end())
+            result.append('\\');
+
+        result.append(character);
+    }
+
+    return result.toString();
+}
+
+// https://urlpattern.spec.whatwg.org/#escape-a-pattern-string
+String escapePatternString(StringView input)
+{
+    ASSERT(input.containsOnlyASCII());
+
+    if (input.is8Bit())
+        return escapePatternStringForCharacters(input.span8());
+
+    return escapePatternStringForCharacters(input.span16());
+}
+
+
+// https://urlpattern.spec.whatwg.org/#is-a-valid-name-code-point
+bool isValidNameCodepoint(UChar codepoint, bool first)
```

Comment:
The second parameter should be a enum class like:
```
enum class IsFirst : bool { No, Yes };
```
It would make the call sites a lot more readable.